subprojects/core/src/main/java/org/gradle/api/internal/collections/FilteredCollection.java

```diff
@@ -156,9 +156,9 @@ public boolean remove(Object o) {
     @Override
     public int size() {
         int i = 0;
-        // TODO this will realize all pending elements
-        for (T o : collection) {
-            if (accept(o)) {
+        // TODO this will realize all pending elements that may be filter out by the filter#spec, still better than realizing everything
```

Comment:
I don't think there is anything we can do about `collection.matching { }` filters as the spec requires a realized element, so this fix is probably as good as we're going to get.

Reply:
We should probably use `iterator()` in `isEmpty()` above as well.